src/rgw/rgw_sal_motr.h

```diff
@@ -40,6 +44,215 @@ class MotrStore;
 #define RGW_MOTR_BUCKET_HD_IDX_NAME   "motr.rgw.bucket.headers"
 //#define RGW_MOTR_BUCKET_ACL_IDX_NAME  "motr.rgw.bucket.acls"
 
+// Notes for Motr FDMI:
+// (1) Currently FDMI can only trigger events for index operation, not object
+//     operation.
+// (2) When an index op event is generated, the FDMI filter picks up the events
+//     matching the pre-defined rules. The FDMI event contains the key/value of
+//     the index operation. Currently, fdmi can only do sub string matching on
+//     the value part, not key part, and can't specify which index to watch,
+//     the filter rule is applied to all indices.
+//     this 
+// (3) The FDMI event is delivered to an FDMI application which defines a
+//     callback function to process FDMI events.
+//
+// Implementation of object watch-notify using Motr FDMI. 
+// (1) An FDMI filter is added (in Motr configuration yaml file) specifying
+//     a matching sub-string (marker).
+// (2) A notifier creates a set of indices which are used to trigger FDMI
+//     events and to pass messages.
+// (3) For an object, hash(obj_name) is used to pick which index to insert
+//     a new key/value record,
+//     key = unique_fid(obj_name),
+//     value = marker (for fdmi filter) + notification msg.
+//     If multiple rgw instances are trying to write the same objects at the same
+//     time, we assume that Motr index is updated atomically.
+// (4) An FDMI event with a special marker is generated and
+//     picked up by the filter and delivered to our fdmi application, watcher.
```

Comment:
How watchers are registered, how m0d knows who is interested in these FDMI events and where to send them?

Reply:
(1) A FDMI filter rule defines a filter id, this id is used as watcher's unique identifier. Each watcher defines a watch_cb() to process FDMI event. A map between filter id and watcher is used to lookup watcher by filter id.

 static map<struct m0_fid, MotrWatcher*, MotrFidComparator> motr_watchers;

(2) Each rgw instance registers a FDMI plugin callback functions, this callback function receives all the FDMI events (may be picked up by different filters) sent to this rgw instance. This callback function is implemented in process_fdmi_record(). filter id is one of the parameters passed to it. Using the filter id, we can find the watcher, then watcher's callback function will be invoked.

(3) FDMI filter rule has a section to define which client nodes to send the FDMI events filtered out by this rule. See the sample fdmi filter rule below, the client index tells which client nodes to send.
fdmi_filters:
  - name: mgw_cache
    node: centos8
    client_index: 0
    substrings: ["rgw.motr.cache.fdmi.marker"]

Reply:
Am I correctly understand that:

1) all FDMI events will be sent to FDMI plugin apps on the client_index nodes (how are the nodes numbered, btw? which client node has index 0, which 1 and so on?);

2) a plugin app (rgw in our case) has to filder out it's own FDMI even by provided filter id.

This all FDMI architecture sounds awkward, to be honest. If I correctly understood it, of course. Why cannot CAS filter the events and send only the ones that the plugin app is interested in? Why to send all of them flooding the cluster network with needless packets?

cc @nikitadanilov, @huanghua78.

Reply:
@andriytk 
Regarding the 2nd point:
2. a plugin app (rgw in our case) has to filder out it's own FDMI even by provided filter id.

Sorry I didn't explain it clearly, let me clarify here:
1. FDMI filter does only pick up and send the events that the plugin app is interested in. FDMI doesn't send all FDMI events.
2. What I meant to explain in my reply is that: a rgw instance may define multiple filters, for each filter, a callback function is registered, that can be easily done in plugin written in c. But when implementing fdmi callback in c++, because of the function type issue, I can't implement the callback function as a member function of a watcher, the workaround is to define a static function as gateway to dispatch fdmi events to watchers. See my comment before  static int process_fdmi_record() in rgw_sal_motr.cc.
when I said, "this callback function receives all the FDMI events (may be picked up by different filters) sent to this rgw instance."
I mean those fdmi events matching the filters interested by the rgw instance ( see MotrWatcher::init_fdmi_plugin()), not every event.

Reply:
I see, it's more clear now. Thanks!